// @nuxtjs/tailwindcss/dist/module.mjs
import { join, relative } from 'path';
import { existsSync } from 'fs';
import { defuArrayFn } from 'defu';
import chalk from 'chalk';
import consola from 'consola';
import { defineNuxtModule, resolvePath, createResolver, requireModule, addTemplate, isNuxt2, installModule, addDevServerHandler } from '@nuxt/kit';

const name = "@nuxtjs/tailwindcss";
const version = "5.1.3";

const defaultTailwindConfig = ({ srcDir }) => ({
  theme: {
    extend: {}
  },
  plugins: [],
  content: [
    `${srcDir}/components/**/*.{vue,js,ts}`,
    `${srcDir}/layouts/**/*.vue`,
    `${srcDir}/pages/**/*.vue`,
    `${srcDir}/composables/**/*.{js,ts}`,
    `${srcDir}/plugins/**/*.{js,ts}`,
    `${srcDir}/App.{js,ts,vue}`,
    `${srcDir}/app.{js,ts,vue}`,
    `${srcDir}/Error.{js,ts,vue}`,
    `${srcDir}/error.{js,ts,vue}`
  ]
});

const logger = consola.withScope("nuxt:tailwindcss");
const module = defineNuxtModule({
  meta: {
    name,
    version,
    configKey: "tailwindcss"
  },
  defaults: (nuxt) => ({
    configPath: "tailwind.config.js",
    cssPath: join(nuxt.options.dir.assets, "css/tailwind.css"),
    config: defaultTailwindConfig(nuxt.options),
    viewer: true,
    exposeConfig: false,
    injectPosition: 0
  }),
  async setup(moduleOptions, nuxt) {
    const configPath = await resolvePath(moduleOptions.configPath);
    const cssPath = await resolvePath(moduleOptions.cssPath, { extensions: [".css", ".sass", ".scss", ".less", ".styl"] });
    const injectPosition = ~~Math.min(moduleOptions.injectPosition, (nuxt.options.css || []).length + 1);
    if (typeof cssPath === "string") {
      if (existsSync(cssPath)) {
        logger.info(`Using Tailwind CSS from ~/${relative(nuxt.options.srcDir, cssPath)}`);
        nuxt.options.css.splice(injectPosition, 0, cssPath);
      } else {
        logger.info("Using default Tailwind CSS file from runtime/tailwind.css");
        const resolver = createResolver(import.meta.url);
        nuxt.options.css.splice(injectPosition, 0, resolver.resolve("runtime/tailwind.css"));
      }
    }
    let tailwindConfig = {};
    if (existsSync(configPath)) {
      tailwindConfig = requireModule(configPath, { clearCache: true });
      logger.info(`Merging Tailwind config from ~/${relative(nuxt.options.srcDir, configPath)}`);
      if (Array.isArray(tailwindConfig.purge)) {
        tailwindConfig.content = tailwindConfig.purge;
      }
    }
    tailwindConfig = defuArrayFn(tailwindConfig, moduleOptions.config);
    if (moduleOptions.exposeConfig) {
      const resolveConfig = await import('tailwindcss/resolveConfig.js').then((r) => r.default || r);
      const resolvedConfig = resolveConfig(tailwindConfig);
      const template = addTemplate({
        filename: "tailwind.config.mjs",
        getContents: () => `export default ${JSON.stringify(resolvedConfig, null, 2)}`
      });
      addTemplate({
        filename: "tailwind.config.d.ts",
        getContents: () => 'declare const config: import("tailwindcss/tailwind-config").TailwindConfig\nexport { config as default }',
        write: true
      });
      nuxt.options.alias["#tailwind-config"] = template.dst;
    }
    if (nuxt.options.dev) {
      nuxt.options.watch.push(configPath);
    }
    await nuxt.callHook("tailwindcss:config", tailwindConfig);
    tailwindConfig._hash = String(Date.now());
    const postcssOptions = nuxt.options.postcss || nuxt.options.build.postcss.postcssOptions || nuxt.options.build.postcss;
    postcssOptions.plugins = postcssOptions.plugins || {};
    postcssOptions.plugins["tailwindcss/nesting"] = postcssOptions.plugins["tailwindcss/nesting"] ?? {};
    postcssOptions.plugins["postcss-custom-properties"] = postcssOptions.plugins["postcss-custom-properties"] ?? {};
    postcssOptions.plugins.tailwindcss = tailwindConfig;
    if (isNuxt2()) {
      await installModule("@nuxt/postcss8");
    }
    if (nuxt.options.dev && moduleOptions.viewer) {
      const route = "/_tailwind";
      const createServer = await import('tailwind-config-viewer/server/index.js').then((r) => r.default || r);
      const { withTrailingSlash, withoutTrailingSlash } = await import('ufo');
      const _viewerDevMiddleware = createServer({ tailwindConfigProvider: () => tailwindConfig, routerPrefix: route }).asMiddleware();
      const viewerDevMiddleware = (req, res) => {
        if (req.originalUrl === withoutTrailingSlash(route)) {
          res.writeHead(301, { Location: withTrailingSlash(req.originalUrl) });
          return res.end();
        }
        _viewerDevMiddleware(req, res);
      };
      addDevServerHandler({ route, handler: viewerDevMiddleware });
      nuxt.hook("listen", (_, listener) => {
        const viewerUrl = `${withoutTrailingSlash(listener.url)}${route}`;
        logger.info(`Tailwind Viewer: ${chalk.underline.yellow(withTrailingSlash(viewerUrl))}`);
      });
    }
  }
});

export { module as default };
